sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java

```diff
@@ -0,0 +1,228 @@
+    }
+  },
+
+  /** BZip compression. */
+  BZIP2(".bz2", ".bz2") {
+    @Override
+    public ReadableByteChannel readDecompressed(ReadableByteChannel channel) throws IOException {
+      return Channels.newChannel(
+          new BZip2CompressorInputStream(Channels.newInputStream(channel), true));
+    }
+
+    @Override
+    public WritableByteChannel writeCompressed(WritableByteChannel channel) throws IOException {
+      return Channels.newChannel(
+          new BZip2CompressorOutputStream(Channels.newOutputStream(channel)));
+    }
+  },
+
+  /** Zip compression. */
+  ZIP(".zip", ".zip") {
+    @Override
+    public ReadableByteChannel readDecompressed(ReadableByteChannel channel) throws IOException {
+      FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
+      return Channels.newChannel(zip);
+    }
+
+    @Override
+    public WritableByteChannel writeCompressed(WritableByteChannel channel) throws IOException {
+      throw new UnsupportedOperationException("Writing ZIP files is currently unsupported");
+    }
+  },
+
+  /** Deflate compression. */
+  DEFLATE(".deflate", ".deflate", ".zlib") {
+    @Override
+    public ReadableByteChannel readDecompressed(ReadableByteChannel channel) throws IOException {
+      return Channels.newChannel(
+          new DeflateCompressorInputStream(Channels.newInputStream(channel)));
+    }
+
+    @Override
+    public WritableByteChannel writeCompressed(WritableByteChannel channel) throws IOException {
+      return Channels.newChannel(
+          new DeflateCompressorOutputStream(Channels.newOutputStream(channel)));
+    }
+  };
+
+  private final String suggestedSuffix;
+  private final List<String> detectedSuffixes;
+
+  Compression(String suggestedSuffix, String... detectedSuffixes) {
+    this.suggestedSuffix = suggestedSuffix;
+    this.detectedSuffixes = Arrays.asList(detectedSuffixes);
+  }
+
+  public String getSuggestedSuffix() {
+    return suggestedSuffix;
+  }
+
+  public boolean matches(String filename) {
+    for (String suffix : detectedSuffixes) {
+      if (filename.toLowerCase().endsWith(suffix)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public boolean isCompressed(String filename) {
+    Compression compression = this;
+    if (compression == AUTO) {
+      compression = detect(filename);
+    }
+    return compression != UNCOMPRESSED;
+  }
+
+  public static Compression detect(String filename) {
+    for (Compression value : values()) {
+      if (value.matches(filename)) {
+        return value;
+      }
+    }
+    return UNCOMPRESSED;
+  }
+
+  public abstract ReadableByteChannel readDecompressed(ReadableByteChannel channel)
+      throws IOException;
+
+  public abstract WritableByteChannel writeCompressed(WritableByteChannel channel)
+      throws IOException;
+
+  /** Extend of {@link ZipInputStream} to automatically read all entries in the zip. */
```

Comment:
<!--new_thread; commit:12876d43a961dd8ff9dba5f4cffcfbe643927eee; resolved:0-->
Suggested class comment:
`Concatenate all {@link ZipInputStream}s contained within the zip file.`

Reply:
Done.